libcnb/src/layer/struct_api/mod.rs

```diff
@@ -0,0 +1,221 @@
+pub(crate) mod handling;
+
+// BuildContext is only used in RustDoc (https://github.com/rust-lang/rust/issues/79542)
+#[allow(unused)]
+use crate::build::BuildContext;
+use crate::layer::shared::{replace_layer_exec_d_programs, replace_layer_sboms, WriteLayerError};
+use crate::layer::LayerError;
+use crate::layer_env::LayerEnv;
+use crate::sbom::Sbom;
+use crate::Buildpack;
+use libcnb_data::generic::GenericMetadata;
+use libcnb_data::layer::LayerName;
+use serde::Serialize;
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::path::{Path, PathBuf};
+
+/// A definition for a cached layer.
+///
+/// Refer to the docs of [`BuildContext::cached_layer`] for usage examples.
+pub struct CachedLayerDefinition<'a, M, MA, IA> {
+    /// Whether the layer is intended for build.
+    pub build: bool,
+    /// Whether the layer is intended for launch.
+    pub launch: bool,
+    /// Callback for when the metadata of a restored layer cannot be parsed as `M`.
+    ///
+    /// Allows replacing the metadata before continuing (i.e. migration to a newer version) or
+    /// deleting the layer.
+    pub invalid_metadata: &'a dyn Fn(&GenericMetadata) -> MA,
+    /// Callback when the layer was restored from cache to validate the contents and metadata.
+    /// Can be used to delete existing cached layers.
+    pub inspect_restored: &'a dyn Fn(&M, &Path) -> IA,
```

Comment:
I feel like there might be a clearer name for `inspect_restored` but I can't quite come up with one. Want to iterate on this and some of the other naming comments in a huddle?

Reply:
Slack call outcome: We were leaning towards something like `invalid_metadata_action` and `restored_layer_action`

Reply:
[f7bc0f1](https://github.com/heroku/libcnb.rs/pull/814/commits/f7bc0f1b540f5f3253c1f411a3c57bf561bd9129)